File: leetcode/2020/april-contest/week2/stack.py
class MinStack:
    def __init__(self):
        self.main = []
        self.mins = []
        self.currentMin = float("inf")

    def push(self, x: int) -> None:
        self.main.append(x)
        if x < self.currentMin:
            self.currentMin = x
        self.mins.append(self.currentMin)

    def pop(self) -> None:
        self.main.pop()
        self.mins.pop()
        if len(self.mins) == 0:
            self.currentMin = float("inf")
        elif self.mins[-1] > self.currentMin:
            self.currentMin = self.mins[-1]

    def top(self) -> int:
        return None if len(self.main) == 0 else self.main[-1]

    def getMin(self) -> int:
        return None if len(self.mins) == 0 else self.mins[-1]

File: leetcode/2020/april-contest/week2/stack.cpp
#include <bits/stdc++.h>
using namespace std;

class MinStack {
private:
  stack<int> mainStack;
  stack<int> minStack;
  int currentMin;

public:
  /** initialize your data structure here. */
  MinStack() { currentMin = numeric_limits<int>::max(); }

  void push(int x) {
    if (x < currentMin) {
      currentMin = x;
    }

    mainStack.push(currentMin);
    mainStack.push(x);
  }

  void pop() {
    if (!mainStack.empty()) {
      mainStack.pop();
      minStack.pop();
    }
    currentMin = minStack.empty() ? numeric_limits<int>::max() : minStack.top();
  }

  int top() { return mainStack.top(); }

  int getMin() { return minStack.top(); }
};

int main() {
  MinStack stk;
  stk.push(-2);
  stk.push(0);
  stk.push(-3);
  cout << stk.getMin();
  stk.pop();
  cout << stk.getMin();
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */
